### ISSUE TITLE
per-file units

### ISSUE THREAD

**Issue by jessfraz** [KittyCAD/modeling-app#3960]
So @Irev-Dev talked about this but say you have a project dir with subdirectories like kcl-samples, you’d want that to be one project with many sub directories, so instead of setting the units based on just the top level directory we should set the units based on the one closest to the file and then traverse backwards to parents.

this is currently how the cli works and this works really well.

this also means I’d say you clone someone else files into yours in a subdirectory it would work for whatever project.toml they pushed to git

so this is like a huge win.

the other thing is and I can open a separate issue but we should split units setting out of project.toml and because you want to commit your units to git but you don’t wanna commit your theme etc. so we need two files. Settings for stuff that you want in git and user specific shit

**Comment by jessfraz** [comment]
Here’s the issue for separating the two https://github.com/KittyCAD/modeling-app/issues/3961

**Comment by Irev-Dev** [comment]
Another thing we talked about somewhat related, putting it here so it doesn't get lost.

When we get to imports and people sharing modules, it makes sense that modules can be unitless (nothing committed for units in the project.toml), so they will default back to the user importing the files base unit.

In this case where you export a unitless bit of configurable geometry (takes input dimensions to update the geometry), but you still need to guarantee the size of some parts of it, for example something that needs to mount on DIN rail, this is when you'll use the `mm()/inch()` functions. i.e. strategically opting into units for sections of the model.

**Comment by jessfraz** [comment]
Also just to note i think we should do https://github.com/KittyCAD/modeling-app/issues/3961 BEFORE we implement this issue, because we dont want all settings to change we literally only care about the ones in git, which currently are units :)

**Comment by jessfraz** [comment]
otherwise your theme could change in sub dirs and that seems like unnessary complexity and who the fuck even wants that, idk supporting that seems like insanity since we already have so many bugs with the theme changing and looking weird, so please dont do that until at least all the other bugs with weird theme thigns are fixed, if you absolutely want to, but i wouldnt do it at all

**Comment by lf94** [comment]
I think @franknoirot removed per-project theming. It's all user theming now.

**Comment by jessfraz** [comment]
oh thank fucking god

**Comment by jessfraz** [comment]
less features I LOVE IT

**Comment by lf94** [comment]
Given I know who you are and how you express yourself I figured :rofl:

**Comment by jessfraz** [comment]
okay i jjust looked tho and the rust side still has those settings for projects so we should clean that up https://github.com/KittyCAD/modeling-app/blob/main/src/wasm-lib/kcl/src/settings/types/project.rs#L64

**Comment by jessfraz** [comment]
since we'd be renaming files and migrating shit anyways its a good change to clean up what actually can change app level vs. project level vs. dir level

**Comment by Irev-Dev** [comment]
> Aside: I think @franknoirot removed per-project themeing. It's all user-level themeing now.

I think we should be careful not add anything that could be considered a "preference" to the project.toml in future, units are needed for the model, theme is not haha.

**Comment by franknoirot** [comment]
Yeah 100% agree @Irev-Dev. I actually [don't think we need any project-level preferences](https://github.com/KittyCAD/modeling-app/issues/3961#issuecomment-2379585975) (other than possibly themeColor if we think that's worth it), and the only project-level (or dir-level) setting we need is units.

**Comment by franknoirot** [comment]
But I think we don't need even that if we implement [proper unit types in KCL](https://github.com/KittyCAD/kcl-experiments/pull/7).

Currently, KCL files cannot fully define concrete geometry on their own. They represent abstract, unit-less geometry until either a unit standard library function (like [`mm()`](https://zoo.dev/docs/kcl/mm)) is applied or an external unit setting is read in.

The fact that our standard library functions take raw number expressions is an implicit external dependency that makes any KCL file not properly stand on its own. This is exemplified by issues like [this one](https://kittycadworkspace.slack.com/archives/C061R186RUZ/p1727386746076009?thread_ts=1727380743.132359&cid=C061R186RUZ), where generated text-to-cad KCL was only correct if it was accompanied by an external unit setting.

If types are a part of the KCL syntax, and therefore the default unit setting was just that, what unit gets written by default when using the UI, then the following becomes true:
- we don't need any project-level settings at all, and therefor don't need to handle any kind of settings hierarchy
- it becomes impossible to receive KCL code that breaks due to mismatched or missing settings
- what is implicit today when you change the "default unit" setting can be made into a safe, explicit command to convert all the units in your file from one unit to another
- we can provide UX to convert an expression of any unit to any other unit, in the code editor and elsewhere

**Comment by jessfraz** [comment]
I dont think it should be required, thats a huge burden in a file where most files are the same unit and we should push people to most files being the same units. I consider this the equivalent of the other unit functions we have which i think i should have made less easy to use _because_ we don't want people uysing them.

Anyways my negatives are:
- most files use the same units everywhere, this is a huge burden for that
- it opens the door for the anti-pattern of mixing units, its a huge context burden to read code with mixed units and actually understand
- we want to make mixing units hard since its a bad pattern, this makes it way too easy

**Comment by jessfraz** [comment]
i personally loive that the car wheel example breaks when the units are off, BECAUSE WE WANT TO FORCE people to not mix units, because if you dont it scales well and you get so many wins. I don't want to make mixing units easier. It is a terrible pattern and no one will ever use code cad if the file is unreadable.

**Comment by jessfraz** [comment]
also for libraries we will want "unitless" consider jordans letter library or the zoo logo, those would be unit less and that would work out of the box today!

PLUS for libraries like a car wheel, you could use the unit functions we have today to even make that unitless, anyways I am pro the paradigms we set up for today because they work great with the future paradigms as well

**Comment by jessfraz** [comment]
anyways I do not want to optimize for the anti-pattern, we need to stop feeling bad about the anti-pattern kinda sucking, BECAUSE IT SHOULD

**Comment by franknoirot** [comment]
> most files use the same units everywhere, this is a huge burden for that
Is the huge burden having `.in` specified after each raw number? That doesn't seem like a huge burden, maybe I'm missing something.

> we want to make mixing units hard since its a bad pattern, this makes it way too easy
Mixed unit parts are incredibly common though; even the category of adapter pipe fittings is massive. In shops that build assemblies, for example, it's really common to have the screw holes in whatever your locality is, but then the important dimensions in whatever your customer's dimensions are. We'd be expecting machine shops with access to metric hardware building for imperial to buy imperial

**Comment by jessfraz** [comment]
yes but if you mix units, its not that we don't support it today, it just takes more effort, which it should since we don't want people doing it unless super necessary

**Comment by jessfraz** [comment]
`.in` is three more characters to type EVERY TIME YOU WRITE SOMETHING thats insane to me

**Comment by franknoirot** [comment]
I don't think mixed units is an anti-pattern, it's a fact of life. There are many many parts that are built in the world and will that use different unit systems. That must be where I'm misunderstanding.

**Comment by jessfraz** [comment]
lets let @JordanNoone and @jgomez720 and @nicboone8 chime in but im super anti doing this
even just the context it takes the read the code has an overhead because usually I can look at code and grok it, but like now im like fuck are the units just going to switch somewhere, like i feel like the other is more readable but maybe im just biased

**Comment by jessfraz** [comment]
c makes no sense and is unreadable, yes you need more characters to understand shit sometimes

**Comment by franknoirot** [comment]
> c makes no sense and is unreadable, yes you need more characters to understand shit sometimes

Yeah it's an extreme example, but I read `angledLine({ length: 2, angle: 30 })` and think "2 of WHAT"?

**Comment by jtran** [comment]
> yes but if you mix units, its not that we don't support it today, it just takes more effort, which it should since we don't want people doing it unless super necessary

I feel like you're not thinking big enough. What I hear you actually saying is that you don't want to degrade the experience when you're only using one kind of unit. I think we can do both.

By adding explicit units, we make the "difficult" case of mixed units easy because the language and tools can help you by providing better error messages (for when you use an angle when a function expects a length) or even automatically converting between inches and millimeters when appropriate.

Have you ever used [uom](https://github.com/iliekturtles/uom) for dealing with measurements in Rust? Compile time checking of units. It's amazing. So hard to go back once you've experienced them.

On the other hand, if everything is unspecified, the tool can't understand what anyone means and can't provide assistance.

**Comment by jessfraz** [comment]
everything is not unspecified tho, we literally know the units from the units file, and we use that at executor time, i guess you are saying to have it at parse time but im unsure what we gain from that since we do both at the same time in the app / cli anyways

**Comment by jessfraz** [comment]
I think this all comes down to what the MEs want tbh, if I was an ME i wouldnt want to type three extra characters everytime and have the mental overhead of reading files with units, but thats just me

**Comment by franknoirot** [comment]
Yeah totally fair, and I see your point. I don't want to optimize for the typing experience first, and I would weight having units of measure be absolutely explicit in one file over the increase in typing friction. I think it has huge implications for code generation.

**Comment by jessfraz** [comment]
heres another idea too, because I dont want to discount your assemblies have multiple units, yes they do , maybe you import something from mcmaster or whatever, BUT individual parts do not

so take the wheel thats actually an assembly, but if the lugs etc where seperate parts it would not be mixed units UNTIL the assembly file

you see where im going so THIS better supports the whole traverse directories thing. or per-file units. WHERE per-file YOU really DONT want to be mixing units UNLESS its an assembly. in which case yes mayeb in assemby files we make it easier to mix units BUT do you even need that... if they are all sepoerate parts with their own units being imported jsut being mated together?

**Comment by jessfraz** [comment]
to even your make the holes in inches to support the inches lugs if you import a varaibel from a file that is inches it would still be inches, so you COULD do that it would work out of the box with the directory stucture and be invisible to the whole units inlined thing, then when you open a file you dont have this awful context switching of units in the file, you are in one unit that entire time, i like not having that mental burden ya know, like oh i open the lug im in inches

**Comment by jessfraz** [comment]
having import would really make it easier to test all these thoeries 😇 i would like to say lets not close this discussion but let us instead wait until we have import so we can experiment and really see the workflow here, does that work, because i get your points, i do, i just want to make sure adding the mental overhead / typing overhead to every file IS A ABSOLUTE MUST first

**Comment by franknoirot** [comment]
> BUT do you even need that... if they are all sepoerate parts with their own units being imported jsut being mated together?

Oh yeah with my "screws in metric, length in imperial" example, that would probably be covered by what you're talking about for the most part, like the screw files would be all-metric, but the diameter of those holes would be metric too (in this story they have access to metric hardware like screws and drill bits, I'm borrowing from [this guy's video](https://youtu.be/qtJ9eHAL0Ns?si=PyH1WZTln1GpIS3L&t=241), but need a part that has critical dimensions), even if the whole outer dimensions of the assembly are in imperial

**Comment by jtran** [comment]
> if you import a variable from a file that is inches it would still be inches

...regardless of the setting of the user who opened it. I think this point is crucial.

**Comment by jessfraz** [comment]
right but the diameter of the holes could be imported from the "imperial file" into a metric file AND IT WOULD WORK thats the coolness of it

**Comment by jessfraz** [comment]
exactly @jtran 

now i really want import to mess around, but yeah maybe we end up realizing thats shit, listen im not always right, but its worth not doing a hiuge overhaul until we know it sucks

**Comment by franknoirot** [comment]
> i like not having that mental burden ya know, like oh i open the lug im in inches

Yeah this is totally fine, and I think having a default unit **setting** and doing the change outlined in this issue is actually not mutually exclusive with having explicit units always in the file. I spoke too soon earlier, apologies. You could enforce the default unit on a per-directory level with these settings, which would simply dictate what the default selected unit would be when users go to perform actions in that directory, _and_ have the file contain explicit units.

**Comment by jessfraz** [comment]
dont apologize i like the discussion it made me see a lot of things  and it makes us more aware going into import / assemblies on how to make it good and problems that might occur

**Comment by franknoirot** [comment]
So this traversal effort would still be fruitful. I just don't think it should be done to work around the fact that our geometry-producing standard library functions can take raw numerical values, forcing us to use an external setting to make the geometry concrete.

**Comment by jessfraz** [comment]
yeah also just to reiterate def push back on work if you see it going nowhere, so im glad you spoke up

**Comment by jessfraz** [comment]
okay also what would be really cool is if in the assembly files or just anytime you import a variable from somewhere when you hover over it, it told you the units, that would be dope and we could def do something like that in the lsp w the data

**Comment by jgomez720** [comment]
I realized that my face is blocking the dimensions in SketchUp. But sketching is asking for things like
`2'36"`, `30"`, `2.32m` with every single dimension. I used SketchUp professional with Spatial for about a year and did not like it. 
https://github.com/user-attachments/assets/31609634-55e0-41ed-be68-b16c9c76c184

**Comment by franknoirot** [comment]
@jgomez720 I'm talking about requiring units in the KCL code, not in the command palette input or any other UI. Your settings would merely choose what units in this dropdown I show in the video gets selected by default. The _KCL code generated by the extrude command_ would require a unit value, but your expression input in the UI would not.

Uploading Screenshare - 2024-09-27 5_27_56 PM-compressed.mp4…

**Comment by jgomez720** [comment]
Yeah my point was more that typing in the .in or .mm would get annoying to me after a while. So this
> I think this all comes down to what the MEs want tbh, if I was an ME i wouldnt want to type three extra characters everytime and have the mental overhead of reading files with units everywhere, but thats just me

is what I agree with. I'm pretty confident that people are not going to want to add the units to every single dimension. I have a habit of writing my units on constants like so:
```js

// Define constants
const mountingHoleDia = .625 // inches
const baseDia = 4.625 // inches
const pipeDia = 1.25 // inches
const thickness = .625 // inches
const totalThickness = 0.813 // inches
const topTotalDiameter = 2.313 // inches
const bottomThickness = 0.06 // inches
const bottomTotalDiameter = 2.5 // inches
const mountingHolePlacementDiameter = 3.5 // inches
const baseThickness = .625 // inches
const topTotalThickness = totalThickness - (bottomThickness+baseThickness) // inches
const holeLocator = baseDia - 8
const nHoles = 4
```
and some people would even think _that_ is too much, let alone every number in the file. 

The UI stuff you showed is tight btw

**Comment by franknoirot** [comment]
Totally hear that, but I really think we should have the KCL be the source of truth rather than leave it up to chance. As it stands today, your comments should really say this:

```
const mountingHoleDia = .625 // inches, hopefully
```

**Comment by franknoirot** [comment]
And when something like text-to-kcl starts consuming that code and producing kcl without a settings.toml associated with it, you get the car-wheel problem. Every tool that uses kcl will need to also understand settings.toml

**Comment by jgomez720** [comment]
The Text-to-KCL point is really good, because giving the user a toml file too would be difficult. 

This was something I did wayyyyyyy back in the day, but maybe this crazy idea actually might be worth looking at? Where we have definitions within the KCL file itself. Forgive my psuedocode, I have no idea what to use for these definitions like `units` or `material`

```js
// Bracket

!units = "in"
!material = "aluminum-6061"

// Define constants
const sigmaAllow = 35000 // psi
const width = 6
const p = 300 // lbs
const factorOfSafety = 1.2 // unitless
const shelfMountL = 5
const wallMountL = 2
```

A point that is also worth discussing is non-dimenions units and non-length units. What do we do with `sigmaAllow`, `p`, and `factorOfSafety` in the above example?

**Comment by jgomez720** [comment]
I still want to pull away from writing the units on every single number. I really don't think people are going to like that. And in other units, are we expecting them to write `const sigmaAllow = 35000.psi` ?

**Comment by franknoirot** [comment]
> I still want to pull away from writing the units on every single number. I really don't think people are going to like that. And in other units, are we expecting them to write const sigmaAllow = 35000.psi ?

Yes. This becomes even more critical when we support other types of units. Imagine your safety checks being wrong because of some other external file!

Numbers would still be valid in this system, so your factor of safety constant would remain just a number, and could be used in expressions that resolve to a dimension.

**Comment by lf94** [comment]
I've read over _everything_ and this is what I'm seeing as meeting all the above criteria.

1. We will always need implicit or "world" / "global" units. 
2. We need a way to transfer these implicit units across **individual KCL files**.
3. We need a way to specify units **at the source code level**.
4. We need a way to specify units **at the project level**.

Why do we need all these ways? Convenience, in my opinion based on what I've read.

1. It has been explicitly stated writing the unit on every number is undesirable / tiring.
2. It has been explicitly stated it reads as noisier.

There is of course familiarity at play here too.

---

To reiterate, we can easily characterize our needs with:

"We need to be able to set units at directory, file, and content levels".

---

What do we need to achieve this?

Well we already have solutions (or partial solutions) to directory and content level units.

We have no solution for file level units. We _need_ a way to put this information there. @jgomez720 has already shown a prototypical example.

So the reality is we are not far from where we want to be. A little re-org needs to happen on those 2 other levels (directory traversal and built-in KCL units as part of the syntax and not functions), and the file level needs a configuration thing.

**Comment by jtran** [comment]
This example code from above:

```javascript
// Define constants
const mountingHoleDia = .625 // inches
const baseDia = 4.625 // inches
const pipeDia = 1.25 // inches
const thickness = .625 // inches
const totalThickness = 0.813 // inches
const topTotalDiameter = 2.313 // inches
const bottomThickness = 0.06 // inches
const bottomTotalDiameter = 2.5 // inches
const mountingHolePlacementDiameter = 3.5 // inches
const baseThickness = .625 // inches
const topTotalThickness = totalThickness - (bottomThickness+baseThickness) // inches
const holeLocator = baseDia - 8
const nHoles = 4
```

...screams out at me that units on every value are _so useful_ that you would go to the trouble to write them in comments, even though the language completely ignores them. Imagine if the tool could actually read all the information in those comments and assist you in your work.

What Frank is trying to say is that we can come up with ways to make it implicit, definitely in the point and click UI, but potentially in the KCL as well. I've definitely thought about a file-level pragma that sets the preferred units. But it gets a bit more complicated if you want to deal with angles, pressure, percentages, etc.

I think we can do unit inference (the same way that type inference is done) so that they can be omitted in many places, and the information is seamlessly reconstructed. But we have to be careful in how we design it. If it's too implicit, it can't be meaningfully inferred.

I really think that this is one of those knee-jerk reactions because you're not familiar with it. But if you've ever actually used it, you'd wonder how anyone gets anything done with measurements without native support.

**Comment by JordanNoone** [comment]
Units should be defined locally per file. There are times you would mix parts with different units in the same assembly. This should be a config setting, +1 to what @jgomez720 wrote above. 

!units = "in"
!material = "aluminum-6061"

or similar 

For me personally I would ideally want to start every file with
import jordan's_config_defaults 

Then override as necessary. In the future this would be for units, tolerances, manufacturing settings (such as for a GLTF slicer, etc...). 

I don't know what is the project.TOML, I don't know what is supposed to be in the project.TOML, I don't know how I'm supposed to know what is supposed to be in the project.TOML, I've never opened it, nor have I ever rev controlled it alongside side the rest of my KCL files that are in github. Before reading this I've been tempted to add it to my .gitignore. If we do use it for something, it needs to be obvious what to the user. 

At this time I don't have a preference on individual lines having units as \\ comments or as some sort of smart units object. If there is one better for Text-to-CAD training, I'd bias towards that.

I elaborate a bit below on "smart units" but I do think this is way out of scope now if not forever as it's a ton of complexity.

The extensibility of a "smart units objects" could be very advantageous to help guarantee the writer keeps unit/dimension consistency between sizing hand calcs. As an example, if they are adding lengths of dissimilar units, that perhaps the output is in the default unit (or some more well thought out behavior), perhaps to errors, perhaps it flags a warning. 

Lots of the annoyance in writing code that sizes engineering features is unit consistency. Analysts will often convert ALL inputs to kg-m-s (kilogram, meter, second) units because they have the best dimensional consistency in units. This is done even if all inputs/outputs need to be in pounds, feet, inches, etc... For example this avoids the "random" *32.2 or /32.2 conversions that show up when dealing with ft-lb-s (feet, pound mass, second) units. There is the common joke/meme/confusion about the slug unit, which is part of the root of the 32.2 factor showing up (https://www.reddit.com/r/EngineeringStudents/comments/j8wl4g/still_have_no_clue_what_a_slug_is_tho/). It's not actually "random", but it often feels that way.

**Comment by jessfraz** [comment]
def dont over-rotate on text-to-cad needing per number units, it doesnt, it could tell us the units per file, and we could generate the settings for it. and we need multi-file text-to-cad for assemblies anyways, so i wouldnt do something just for text-to-cad if its the wrong paradigm for the app. In jordans example above i would consider what he wants the new and improved `git pushable` `config.toml` or whatever we decide upon and we can educate the user its just a key value store of their preferences, and they can change it per directory or whatever heirarchy they want. we will take the defaults from the closest we find, and traverse upwards

such that 

/my/dir/file/thing.kcl
/my/dir/file/other.kcl
/my/dir/file/config.toml

thing.kcl and other.kcl uses config.toml

/my/dir/file/thing.kcl
/my/dir/file/other.kcl
/my/dir/file/config.toml
/my/dir/file/nested/main.kcl

main.kcl uses the parents config.toml since there is none in that dir

etc etc

anyways, this is a paradigm that most programming languages also use and thats why its my preference, but we will see when we get imports and mess around with files where yiou are importing things from different units just how bad or good it is

just lets not design weird paradigms for a user just because we think text-to-cad needs it, text-to-cad can understand multuple files and it can understand per-unit files. 

i personally REALLY like not having the cognitive overload of changing units IN A SINGLE part, in an assembly thats different but we can integrate into the lsp on hover stuff like imported variables units and lessen the cognitive load. just my preference.

**Comment by jessfraz** [comment]
i'd also argue the file does not need units in it, if you can just move it to a nested dir with a different config. for simpliciy sake i say we start there, and see our limitations before we over complicate matters. this is already working great for kcl-samples.

**Comment by jessfraz** [comment]
once we add something we can never remove it, we can never get the time back etc etc so im just trying to keep this iterative, try it out, find limitations go from there

**Comment by nicboone8** [comment]
(Trying my best to make everyone here equally mad at me)

Managing multiple unit types within the same part is important, and it's going to keep coming up as we gain more and more users. Not just within the same assembly.

Having to go in and manually type *mm(), *in(), etc, to declare that a dimension differs from the base unit is a fair punishment for choosing to mix units. Having to type them in for every single number is tedious - even if it's done in smart ways through the UX there would still be instances where you have to go back in the code and change things, which could be annoying if there's a bunch of units everywhere. Maybe the unit inference scheme solves this

Josh's example of base units and other part properties living in the first lines of the kcl script is what makes the most sense to me. A decent number of MEs are going to have a difficult time dealing with multiple file types for one part. And people are always going to try and just copy and paste kcl from god knows where. Josh's idea avoids every version of the car wheel problem

**Comment by jessfraz** [comment]
I just want to say _if_ we embed units and other config values in the top of the file I feel VERY STRONGLY it needs to be the same data structure as the config file itself.

this will help with maintainability, it uses the same config file parsing logic and passing said structure to the executor would be the same

if we add more values to the config file this would work both places without any need for custom code

it’s only one config file syntax people need to learn

not doing this would over complicate what the user needs to learn and our code, and adding a new config value each time should not be some one off parsing logic behavior it should just reuse the config parsing we already have.

**Comment by lf94** [comment]
> I feel VERY STRONGLY it needs to be the same data structure as the config file itself.

Are you suggesting one of the two? If not, well, my imagination doesn't extend far enough for this one...

1. TOML inside KCL files?
2. A single KCL object is exported from a .kcl file and imported into other files?... (KCLO? KSON? lol.) Does the imported KCL object need to be assigned to a special global?

**Comment by jessfraz** [comment]
1. Which looks a lot like how rust is proposing it as well.

**Comment by jessfraz** [comment]
This https://internals.rust-lang.org/t/pre-rfc-cargo-script-for-everyone/18639

**Comment by lf94** [comment]
What a great resource! I will read through it all and report back. It is hard to image embedding TOML in a comment at the beginning of source files is an acceptable solution as it goes directly against Rust's ethos of being robust...

---

20 minutes later _spongebob french narrator voice_

---

Ok. The motivation seems to be there is **no choice that it is TOML**. Since all prior cargo data is TOML, there is no choice other than TOML to be embedded.

Additionally they are proposing a separate file extension, something like `.rss`, for "Rust script source". It is intended to be a self-contained Rust project pretty much. For us that's a problem. We want KCL to be usable with other KCL even with their own environment settings.

---

It kind of looks like that proposal was abandoned. https://github.com/epage/cargo-script-mvs/compare/1bf0dfec25b7543d03ab53765b39c1a40d2f32ef...main

Both the discussion and commits have stopped 2 years ago.

---

All in all it's hard for me to conclude if we want to execute on the idea of mixing languages (not that it matters since it's not my decision to make). It feels wrong when we have a choice right now to use KCL native data structures to express environment information. The Cargo crowd doesn't. I'm not really convinced by what I've just taken in... We don't have the same problems to deal with.

---

What I would **love** is to ask Ed Page what he thinks about our situation :) Seems like an absolute pro in this space, as they are the maintainers of `winnow`, `toml` and all the cargo stuff.

[@epage@hachyderm.io](https://hachyderm.io/@epage)
[@orenpage](https://twitter.com/orenpage)

---

I'm super duper convinced we should focus on `import` like immediately after whatever language developments are going on right now.

**Comment by jessfraz** [comment]
Okay I have thought about this at length and I changed my mind, I'll explain.

TLDR; I am now PRO per-file units (& other git-pushable config settings) under the condition THERE IS NO OTHER WAY TO SET THEM

So basically my fear with per-file units was people are all just going to use that, between our samples, text-to-cad, etc theyd just use that and they'd be pervasive and no one would use any other setting, the other settings wouldnt work well since no one uses them and its be this weird code path we rarely hit and be shit. Well so then its like lets just get rid of every other setting and just make it per file or not at all.

My constraints were:

- preferably there should only be one place that stores your git-pushable settings, lessens the cognitive load
- if we must have 2 places for git-pushable settings THEY MUST USE THE SAME SYNTAX
- you should be able to open a file anywhere and it should just work
- kcl-samples the repo should work at the root, meaning like it should be a "project" even tho i really want to get away from projects but whatever, today you cant just open kcl-samples at the root and have it work you need to point your whole app there and its fucking stupid, this is why people keep copy pasting into those files and no one is using git and opening the directory in the app and testing the actual fucking workflow we are trying to create
- you should be able to import/use another kcl file and it keeps the units for that file or is unitless
- assemblies should be able to import files of various units and keep their units
- josh should stop commenting on every const that its in inches, it should be easily apparent what the file is in
- scripts should be easily sharable with a single file if it imports nothing

So 1. this solves all those. 2. it lessens the cognitive load there is only one place for units settings its in the fucking file and users will know if outside the app to just look there, 3. its at the top of the file so josh wouldn't need to comment those variables anymore since its right fucking there

So my proposal is this:

- we add it to the top of every file, if unspecified we consider it unitless if its being imported or mm if not (which is the default today), you can also explicitly set a file as unitless/mm/etc
- as a migration path on fmt we move the git-configurable settings from the project.toml to the top of the file

as far as syntax goes i have the following constraints but other than that im like whatever:

- dont make it fugly
- it should parse easily into a struct we can add more things to EASILY as in you dont modify the parser every time you only add something to the struct and it just fucking works
- it should self document itself like everything fucking else we have
- it will come after the shebang if there is one
- lets not reinvent the wheeel either make it look like existing kcl or something users can just get.

anyways, these are now my thoughts.

**Comment by jessfraz** [comment]
oh the other win is, this becomes part of the programs AST, making use/imports/assemblies so much easier

**Comment by jtran** [comment]
After a discussion with @lf94, this relieves many of my concerns 👍 

I'm also PRO per-file units, no matter how we do it.

Minor preference for using a KCL object, like @lf94 suggested. The GUI will write it, so most users won't care. But it has lots of other benefits.

**Comment by nicboone8** [comment]
Is it worth supporting unitless kcl files, vs just defaulting everything unspecified to mm? Setting a part to unitless and working with it would feel weird to me, but I'm not opposed to having that functionality if it makes it easier for things like import/export

**Comment by jessfraz** [comment]
@nicboone8 to your point, some libraries will be unitless and such and thats totally fine.

**Comment by jessfraz** [comment]
Okay the rust side is done. I think we need some help on the front-end but i thought id bullet point some stuff.

- [ ] setting the units in the gizmo should change the code for the file
- [ ] the gizmo should show the units from the file
- [ ] remove setting the units in the app / project tomls
- [ ] if you are in a inches file and you start a new file, it should default to inches, this will save the inches nerds time (ie new files should default to whatever the last file opened was), if no last file was opened it can be mm (or i guess we could use the app setting defaults but id love to nuke that to ease complexity) up to whoever implements this
- [ ] i dont think we should overthink backwards compatibility with the project.toml but ill leave that to the implementer, whatever is easier

**Comment by Irev-Dev** [comment]
This is V1 stuff right? because I dare say we don't want to deal with this change post v1

I can see the "default to last file" thing still being frustrating for inchs MEs who happen to open the odd metric file.

**Comment by nicboone8** [comment]
@Irev-Dev I had the same concern initially, but I think I would end up preferring that vs having to select units for every new part. As long as the units are obvious and easy to change, I could get used to it pretty quickly. CC @jgomez720 @nickmccleery in case I'm out to lunch on this

**Comment by jessfraz** [comment]
yes this is v1, @jgomez720 has already started updating the samples as with https://github.com/KittyCAD/modeling-app/pull/5178 we are fully unblocked on this working in kcl, it will work in the app too just the gizmo will be inaccurate, mostly i was pressuring josh since this fixes the wheel

**Comment by jessfraz** [comment]
ya mostly importantly we just need the gizmo to work, and to reset the project.toml / app.tomls so we can eventually remove it, as far as everything else ill leave to you i was just trying to lessen any complexity

**Comment by jessfraz** [comment]
we could keep the units in the app.toml and basically its meaning jsut becomes "always make new files these units" that works as well versus the use the last one opened

**Comment by jessfraz** [comment]
we just need to ensure we are _NOT_ using it for sending to the engine, it merely is effecting new files

**Comment by franknoirot** [comment]
I like that, was going to chime in to say that having a user-level setting for what the default units that get written to new files should be makes sense, it's a setting in all traditional CAD packages.

**Comment by franknoirot** [comment]
I'm a little fuzzy on how I find if there is a units attribute present in the file, both to read for the gizmo's state and to replace with a new unit. I'd guess that doesn't appear in the AST right? I could do a naive string search but that seems smelly

**Comment by jessfraz** [comment]
cc @nrc i thought the data was in the ast? I think its an annotation or something? This might be a good use case for a small API to allow them to modify an ast from the rust side like we've been talking about cc @jtran @adamchalmers as well

**Comment by jtran** [comment]
Jess is right in that it is in the AST.

https://github.com/KittyCAD/modeling-app/pull/5178/files#diff-07ca84911a11c51ce4c10204cab78c471ae124e40bb99556022a7b220d093551R146

+1 for making a Rust API to change it.

**Comment by franknoirot** [comment]
Oh very cool. Yeah I'd prefer an API for messing with that to protect me from myself

**Comment by jessfraz** [comment]
@jtran im happy to do the api side but i havent been marinating on this so if you had a visioin here im happy to do that

**Comment by jtran** [comment]
Eventually, we don't want to expose the AST to TS at all. We've considered a more general refactoring API, but in the short- to medium-term, I think we should make specialized one-off functions for doing things like this. The function would take a KCL string and the new default units, and returns the new KCL string.

We have `src/wasm-lib/kcl/src/parsing/ast/modify.rs`, but it's dead code.

**Comment by jessfraz** [comment]
lol i wrote that file originally, okay @franknoirot ill try and provide this today after calls

**Comment by jessfraz** [comment]
ill also make a function that takes a kcl-string and return the current units

**Comment by jessfraz** [comment]
i actually feel like we should use the ast for that (getting the current units) for the time being otherwise we are going to add in a lot more parsing thats unnessary

**Comment by jessfraz** [comment]
we can probably close out this bug at the same time https://github.com/KittyCAD/modeling-app/issues/5155 get a 2 for 1

**Comment by nicboone8** [comment]
Maybe 3 for 1 #4833

**Comment by franknoirot** [comment]
> i actually feel like we should use the ast for that (getting the current units) for the time being otherwise we are going to add in a lot more parsing thats unnessary

Cool yeah [in my branch here](https://github.com/KittyCAD/modeling-app/compare/main...franknoirot/3960/file-units-ui) I've made a utility in `queryAst` that tries to get the settings annotation during `executeAst` and save it onto `KclManager`, does that seem reasonable?

**Comment by jessfraz** [comment]
seems okay to me!

**Comment by franknoirot** [comment]
@jtran and I are going to close this as complete!

Here is how the settings work now:
1. A setting, available on the user- and project-level, determines only what per-file annotation gets written to new files by default.
    - We have discussed getting rid of the project-level setting, but it provides user utility and takes less work to leave in. If a user is working on a project that has a different policy than they would normally want; like a contractor that normally works in metric (set in as a user-level setting) contributing to a US project in imperial (set as a project-level setting).
2. A settings annotation is the only way to change how a file's unitless lengths are interpreted. If this annotation is not present, all unitless lengths are interpreted as being in `mm`.
3. Any individual length values can have explicit inline units, opting out of the default provided by the per-file annotation.
    - Currently this is done with unit functions like [`inch(5)`](https://zoo.dev/docs/kcl/inch), but the more robust approach using Unit of Measure suffixes is a WIP such as `extrude(sketch001, length = 5in)`. 

There are improvements to make around when we should adjust the setting on the user's behalf, such as this example from @jtran, but we agree these should be broken out into smaller issues for post-V1:

> But in that scenario, say your user setting is mm and you create a new project. The first file you set to in. Save. Git commit. Next week when you return to the project and create a new file, it's going to create it in mm since you didn't set your project setting. It just seems like the app could be a little smarter there.

@jtran, anything else on your mind about this one or do you want to do the honor of smashing that purple button?
